src/tribler/core/components/database/db/layers/tests/test_user_activity_layer.py

```diff
@@ -0,0 +1,201 @@
+from tribler.core.components.database.db.layers.user_activity_layer import UserActivityLayer
+from tribler.core.components.user_activity.types import InfoHash
+from tribler.core.utilities.pony_utils import TrackedDatabase
+
+
+@fixture(name="layer")
+def fixture_activity_layer() -> Generator[UserActivityLayer, None, None]:
+    database = TrackedDatabase()
+    database.bind(provider="sqlite", filename=":memory:")
+    ual = UserActivityLayer(database)
+    database.generate_mapping(create_tables=True)
+    yield ual
+    database.disconnect()
+
+
+def float_equals(a: float, b: float) -> bool:
+    return round(a, 5) == round(b, 5)
+
+
+def test_store_no_losers(layer: UserActivityLayer) -> None:
+    """
+    Test that queries can be stored and retrieved.
+    """
+    layer.store("test query", InfoHash(b"\x00" * 20), set())
+
+    with db_session():
+        queries = layer.Query.select()[:]
+
+        assert len(queries) == 1
+        assert queries[0].query == "test query"
+        assert len(queries[0].infohashes) == 1
+        assert list(queries[0].infohashes)[0].infohash == b"\x00" * 20
+        assert float_equals(list(queries[0].infohashes)[0].preference, 1.0)
+
+
+def test_store_with_loser(layer: UserActivityLayer) -> None:
+    """
+    Test that queries with a loser can be stored and retrieved.
+    """
+    layer.store("test query", InfoHash(b"\x00" * 20), {InfoHash(b"\x01" * 20)})
+
+    with db_session():
+        queries = layer.Query.select()[:]
+        winner, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x00" * 20)[:]
+        loser, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x01" * 20)[:]
+
+        assert len(queries) == 1
+        assert queries[0].query == "test query"
+        assert float_equals(winner.preference, 1.0)
+        assert float_equals(loser.preference, 0.0)
+
+
+def test_store_with_losers(layer: UserActivityLayer) -> None:
+    """
+    Test that queries with multiple losers can be stored and retrieved.
+    """
+    layer.store("test query", InfoHash(b"\x00" * 20), {InfoHash(b"\x01" * 20),
+                                                       InfoHash(b"\x02" * 20),
+                                                       InfoHash(b"\x03" * 20)})
+
+    with db_session():
+        queries = layer.Query.select()[:]
+        winner, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x00" * 20)[:]
+        loser_1, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x01" * 20)[:]
+        loser_2, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x02" * 20)[:]
+        loser_3, = layer.InfohashPreference.select(lambda x: x.infohash == b"\x03" * 20)[:]
+
+        assert len(queries) == 1
+        assert queries[0].query == "test query"
+        assert float_equals(winner.preference, 1.0)
+        assert float_equals(loser_1.preference, 0.0)
+        assert float_equals(loser_2.preference, 0.0)
+        assert float_equals(loser_3.preference, 0.0)
```

Comment:
More compact yet correct version of the same logic. It uses `select` where it is intended to query a list and `get` when it is supposed to be just a single entity.

```suggestion
        test_query = layer.Query.get()
        winner = layer.InfohashPreference.get(lambda x: x.infohash == b"\x00" * 20)
        losers = list(layer.InfohashPreference.select(lambda x: x.infohash != b"\x00" * 20))

    assert test_query.query == "test query"
    assert float_equals(winner.preference, 1.0)
    assert len(losers) == 3
    assert all(float_equals(ls.preference, 0.0) for ls in losers)
```